### PR TITLE
GPTNeoXForQuestionAnswering

### PR DESCRIPTION
**Pull request by peter-sk**
# What does this PR do?

Adds GPTNeoXForQuestionAnswering.

Includes #23030 and #23057.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

@ArthurZucker
@younesbelkada

### CODE REVIEW

**Comment by peter-sk** [comment]
@sgugger @younesbelkada as Arthur is on holdiays 👍

**Review by younesbelkada — COMMENTED** [review]
Hi @peter-sk 
Thanks so much for your dedication! 
It seems that you have copied some code from the other PRs : https://github.com/huggingface/transformers/pull/23030 / https://github.com/huggingface/transformers/pull/23057 / i.e. a bad rebase seemed to happen here. Can you please double check and revert the unneeded changes, thus just keep the changes relevant to `GPTNeoX`?
Thanks!

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 1]
```diff
@@ -974,6 +974,11 @@ def __init__(self, config):
         self.gpt_neox = GPTNeoXModel(config)
         self.qa_outputs = nn.Linear(config.hidden_size, 2)
 
+        # Model parallel
+        self.model_parallel = False
+        self.device_map = None
+        self.gradient_checkpointing = False
```
Comment (sgugger) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:980]:
```suggestion
```
This part should be deleted, this model does not implement the old model parallel

**Comment by peter-sk** [comment]
@younesbelkada @amyeroberts this one is ready for review :-)

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 2]
```diff
@@ -955,3 +956,103 @@ def forward(
             hidden_states=outputs.hidden_states,
             attentions=outputs.attentions,
         )
+
+
+@add_start_docstrings(
+    """
+    The GPT-Neo Model transformer with a span classification head on top for extractive question-answering tasks like
```
Comment (amyeroberts) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:963]:
```suggestion
    The GPT-NeoX Model transformer with a span classification head on top for extractive question-answering tasks like
```

**Review by amyeroberts — APPROVED** [review]
Thanks for adding!

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 3]
```diff
@@ -1670,9 +1669,9 @@ def forward(
         if start_positions is not None and end_positions is not None:
             # If we are on multi-GPU, split add a dimension
             if len(start_positions.size()) > 1:
-                start_positions = start_positions.squeeze(-1)
+                start_positions = start_positions.squeeze(-1).to(start_logits.device)
             if len(end_positions.size()) > 1:
-                end_positions = end_positions.squeeze(-1)
+                end_positions = end_positions.squeeze(-1).to(end_logits.device)
```
Comment (younesbelkada) [inline src/transformers/models/gpt2/modeling_gpt2.py:1674]: Maybe this can live in another PR that will fix that for GPT2

**Review by younesbelkada — COMMENTED** [review]
Looks great! Left one comment, I think we should leave the modifications with respect to GPT2 in another PR

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 4]
```diff
@@ -52,7 +52,6 @@
 logger = logging.get_logger(__name__)
 
 _CHECKPOINT_FOR_DOC = "gpt2"
-_REAL_CHECKPOINT_FOR_DOC = "gpt2"
```
Comment (younesbelkada) [inline src/transformers/models/gpt2/modeling_gpt2.py:55]: Can you move that as well?

**Review by younesbelkada — COMMENTED** [review]
Awesome! One last comment and we should be good to go!

**Review by younesbelkada — APPROVED** [review]
Thank you for this!

File: src/transformers/models/gpt2/modeling_gpt2.py (hunk 4 above)
Comment (younesbelkada) [reply src/transformers/models/gpt2/modeling_gpt2.py:55]: Maybe it's fine as this is quite minor

**Review by sgugger — APPROVED** [review]
Thanks for yet another new model :-)

**Pull request merged**